src/Psalm/Internal/Type/Comparator/ScalarTypeComparator.php

```diff
@@ -55,6 +61,44 @@ public static function isContainedBy(
         bool $allow_float_int_equality = true,
         ?TypeComparisonResult $atomic_comparison_result = null
     ): bool {
+        // TODO: Remove this if statement once all container types have been converted to use this pattern.
+        if ($container_type_part instanceof TArrayKey
+            || $container_type_part instanceof TNumeric
+            || $container_type_part instanceof TScalar
+            || $container_type_part instanceof TBool
+            || $container_type_part instanceof TFloat
+        ) {
+            if ($container_type_part instanceof TArrayKey) {
+                $result = self::isContainedByArrayKey($input_type_part, $container_type_part, $codebase);
+            } elseif ($container_type_part instanceof TNumeric) {
+                $result = self::isContainedByNumeric($input_type_part, $container_type_part);
+            } elseif ($container_type_part instanceof TScalar) {
+                $result = self::isContainedByScalar($input_type_part, $container_type_part);
+            } elseif ($container_type_part instanceof TBool) {
+                $result = self::isContainedByBool($input_type_part, $container_type_part);
+            } else {
+                $result = self::isContainedByFloat($input_type_part, $container_type_part, $allow_float_int_equality);
+            }
+            if ($result) {
+                return true;
+            } else {
+                if ($atomic_comparison_result) {
+                    // The type is coerced if the container type is contained by the input type (the opposite).
```

Comment:
Are you confident about this rule working in every case? I would have like to have TString and TInt comparison in this PR also because those are the most complex and I fear some generalization here could be a problem

Reply:
I cannot think of another way of defining it. It is when you try to put a bigger type into a smaller type.

In the mean time, I'll work on adding TInt support.

Reply:
> It is when you try to put a bigger type into a smaller type.

I agree in the general case but I wonder if there are some edge cases. Think like having a `numeric-string` being pushed where `lowercase-string` is expected. There's not a big or small type in this case, but it should maybe be a scalar coercion still?

Also pushing `int` where `positive-int` is expected would be a scalar coercion, but based on the same definition, pushing a `int<-1, 3>` into a `int<0, 4>` would not even though it's clearly the same kind of operation with restricted range

I just want to be sure we won't merge a partial work for which we will find fatal flaws later